src/storage/src/render/upsert.rs

```diff
@@ -277,8 +277,41 @@ where
         );
         let mut events = vec![];
         let mut snapshot_upper = Antichain::from_elem(Timestamp::minimum());
+
+        let mut stash = vec![];
+        let mut input_upper = Antichain::from_elem(Timestamp::minimum());
+
         while !PartialOrder::less_equal(&resume_upper, &snapshot_upper) {
-            match previous.next_mut().await {
+            let previous_event = tokio::select! {
+                // Note that these are both cancel-safe. The reason we drain the `input` is to
+                // ensure the `output_frontier` (and therefore flow control on `previous`) make
+                // progress.
+                previous_event = previous.next_mut() => {
+                    previous_event
+                }
+                input_event = input.next_mut() => {
+                    match input_event {
+                        Some(AsyncEvent::Data(_cap, data)) => {
+                            if PartialOrder::less_equal(&input_upper, &resume_upper) {
+                                data.retain(|(_, ts, _)| resume_upper.less_equal(ts));
+                            }
+
+                            stash.extend(data.drain(..).map(|((key, value, order), time, diff)| {
+                                assert!(diff > 0, "invalid upsert input");
```

Comment:
I'd be nice to find a way to not have repetition here. It's fine to merge this as is but if there something better we can do it would be amazing.

Reply:
I can make this an internal function, probably?

Reply:
done!